File: algorithms/sorting/go/quicksort.go
package main

import "fmt"

func quickSort(arr []int, low, high int) {
	if low < high {
		pivotIndex := partition(arr, low, high)

		// Recursively sort left and right of the pivot
		quickSort(arr, low, pivotIndex-1)
		quickSort(arr, pivotIndex+1, high)
	}
}

func partition(arr []int, low, high int) int {
	// Implemented where pivot element is last element
	pivot := arr[high]
	i := low - 1 // index of smaller element

	for j := low; j < high; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}

	// Swap pivot to correct position
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

func main() {
	arr := []int{4, 1, 47, 64, 23, 6, 37,69}

	quickSort(arr, 0, len(arr)-1)
	fmt.Println("Sorted array: ", arr)
}

File: algorithms/sorting/quick-sort/QuickSort.py
import random


class Sorting:
    @staticmethod
    def quick_sort_inplace(arr, left, right):
        if left >= right:
            return

        pivot_index = random.randint(left, right)
        pivot = arr[pivot_index]

        arr[pivot_index], arr[right] = arr[right], arr[pivot_index]

        i = left
        for j in range(left, right):
            if arr[j] <= pivot:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1

        arr[i], arr[right] = arr[right], arr[i]

        Sorting.quick_sort_inplace(arr, left, i - 1)
        Sorting.quick_sort_inplace(arr, i + 1, right)
